[shard] use scatter in shard_parameter API

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #72160

This PR switches `shard_parameter` API to use `dist.scatter` instead of `dist.broadcast`. Instead of sending the whole tensor to each rank, we split the tensor beforehand, and only send the part needed to the corresponding rank, which greatly reduce the communication overhead.

Differential Revision: [D33933419](https://our.internmc.facebook.com/intern/diff/D33933419/)

Commit:
[shard] use scatter in shard_parameter API

This PR switches `shard_parameter` API to use `dist.scatter` instead of `dist.broadcast`. Instead of sending the whole tensor to each rank, we split the tensor beforehand, and only send the part needed to the corresponding rank, which greatly reduce the communication overhead.

Differential Revision: [D33933419](https://our.internmc.facebook.com/intern/diff/D33933419/)

[ghstack-poisoned]

Commit:
Update on "[shard] use scatter in shard_parameter API"


This PR switches `shard_parameter` API to use `dist.scatter` instead of `dist.broadcast`. Instead of sending the whole tensor to each rank, we split the tensor beforehand, and only send the part needed to the corresponding rank, which greatly reduce the communication overhead.

Differential Revision: [D33933419](https://our.internmc.facebook.com/intern/diff/D33933419/)

[ghstack-poisoned]

Commit:
Update on "[shard] use scatter in shard_parameter API"


This PR switches `shard_parameter` API to use `dist.scatter` instead of `dist.broadcast`. Instead of sending the whole tensor to each rank, we split the tensor beforehand, and only send the part needed to the corresponding rank, which greatly reduce the communication overhead.

Differential Revision: [D33933419](https://our.internmc.facebook.com/intern/diff/D33933419/)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2326,11 +2326,6 @@ def scatter(tensor, scatter_list=None, src=0, group=None, async_op=False):
         input_tensors = [scatter_list]
         output_tensors = [tensor]
     else:
-        if scatter_list:
-            raise ValueError(
-                "Argument ``scatter_list`` must NOT be specified "
-                "on non-source ranks."
-            )
```
Is this an unintended change?

wanchaol replied:
This is intended change actually, it's to match the behavior in `gather` where `dist.gather` allow user to pass in gather_list on non-dst ranks, but we internally treat it as empty list https://github.com/pytorch/pytorch/blob/defde3bb04e2c98765dbee2d214b024de20626d0/torch/distributed/distributed_c10d.py#L2253

I think the `dist.gather` behavior make more sense, user can call the collective on all ranks SPMD, and inside `dist.gather` we ignore the gather/scatter_list on non-dst/src ranks. Otherwise, user have to write conditional code everytime to perform gather/scatter, i.e. for scatter it must be:
```
if current_rank == src_rank:
    dist.scatter(local_tensor, scatter_list=tensors_to_scatter, src=src_rank, group=pg)
else:
    dist.scatter(local_tensor, scatter_list=None, src=src_rank, group=pg)    
```
which might be very in-convenient for user. Our low level primitive: `pg.gather/scatter` enforce the non-dst/src ranks must pass None.

fduwjj replied:
Is it possible to make this logic (making scatter/gather list to be None for non_src_rank) centralized somewhere in Python? Since we have enforced this in CPP.

wanchaol replied:
it was embedded in `dist.gather` https://github.com/pytorch/pytorch/blob/c1a4714bf1195faa1913541b190e8019f9f5dac5/torch/distributed/distributed_c10d.py#L2261 and `dist.scatter`, https://github.com/pytorch/pytorch/blob/c1a4714bf1195faa1913541b190e8019f9f5dac5/torch/distributed/distributed_c10d.py#L2334, Im just remove the error raising, are you suggesting we have a util function to deal with that? I guess that's a small piece of code only for gather/scatter, maybe we don't need to generalize it out?

pritamdamania87 replied:
The thing is that our docs clearly say it should not be specified on src_rank:

```
        scatter_list (list[Tensor]): List of tensors to scatter (default is
            None, must be specified on the source rank)
```

So this change would be confusing for users where the docs say something whereas the actual behavior is different.

wanchaol replied:
While both `scatter` and `gather` doc does not mention that `scatter_list` "must ONLY be specified on the source rank", so I guess this part is a bit confusing to the user. Anyways, I think it should be a separate discussion around consistency of our c10d APIs, not related to this PR specifically. I just updated the PR to make sure we are passing None on non-dst/src rank, and remove this unintentional change.

wanchaol replied:
created a issue to track this conversation https://github.com/pytorch/pytorch/issues/74323

pritamdamania87 reviewed (commented):
LGTM, just had one comment regarding the change in `distributed_c10d.py`.

fduwjj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -453,31 +455,36 @@ def shard_parameter(
             placement=placement,
         )
         shards_metadata.append(shard_metadata)
+        # Reshape to get shard for each rank and we don't want autograd
+        # recording here for the narrow op and 'tensor_to_scatter' should be a
+        # leaf variable in the autograd graph.
+        tensor_to_scatter = tensor.narrow(
+            sharding_spec.dim,  # type: ignore[arg-type]
+            shard_metadata.shard_offsets[sharding_spec.dim],  # type: ignore[union-attr, arg-type, index]
+            shard_metadata.shard_sizes[sharding_spec.dim],  # type: ignore[union-attr, index]
+        ).clone().detach().contiguous()
+        tensors_to_scatter.append(
+            tensor_to_scatter
+        )
```
Once your new sharding spec extension PR is ready, I guess the logic here can reuse what's been written in the shard api?

wanchaol replied:
yeah agreed, this might just be using `ChunkShardingSpec.shard`

fduwjj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -453,31 +455,36 @@ def shard_parameter(
 
         current_offsets[sharding_spec.dim] += chunked_dim_size  # type: ignore[index]
 
     # Scatter the shards (use broadcast since NCCL doesn't support scatter, this is very inefficient).
```
Nit: You can remove this comment now 😊

fduwjj reviewed (commented):
We can get rid of `broadcast` finally. Does GH support diff review meme?

pritamdamania87 reviewed (commented):
I guess this PR needs to be rebased on top of https://github.com/pytorch/pytorch/pull/72130 or vice-versa?

wanchaol commented:
> 

yep, plan to rebase soon (hopefully can land that PR then rebase to master directly)

Commit:
Update on "[shard] use scatter in shard_parameter API"


This PR switches `shard_parameter` API to use `dist.scatter` instead of `dist.broadcast`. Instead of sending the whole tensor to each rank, we split the tensor beforehand, and only send the part needed to the corresponding rank, which greatly reduce the communication overhead.

Differential Revision: [D33933419](https://our.internmc.facebook.com/intern/diff/D33933419/)

[ghstack-poisoned]

Commit:
Update on "[shard] use scatter in shard_parameter API"


This PR switches `shard_parameter` API to use `dist.scatter` instead of `dist.broadcast`. Instead of sending the whole tensor to each rank, we split the tensor beforehand, and only send the part needed to the corresponding rank, which greatly reduce the communication overhead.

Differential Revision: [D33933419](https://our.internmc.facebook.com/intern/diff/D33933419/)

[ghstack-poisoned]

pritamdamania87 reviewed (changes requested):
Requesting changes since I think scatter behavior should align with our docs

Commit:
Update on "[shard] use scatter in shard_parameter API"


This PR switches `shard_parameter` API to use `dist.scatter` instead of `dist.broadcast`. Instead of sending the whole tensor to each rank, we split the tensor beforehand, and only send the part needed to the corresponding rank, which greatly reduce the communication overhead.

Differential Revision: [D33933419](https://our.internmc.facebook.com/intern/diff/D33933419/)

[ghstack-poisoned]

Commit:
Update on "[shard] use scatter in shard_parameter API"


This PR switches `shard_parameter` API to use `dist.scatter` instead of `dist.broadcast`. Instead of sending the whole tensor to each rank, we split the tensor beforehand, and only send the part needed to the corresponding rank, which greatly reduce the communication overhead.

Differential Revision: [D33933419](https://our.internmc.facebook.com/intern/diff/D33933419/)

[ghstack-poisoned]

Commit:
Update on "[shard] use scatter in shard_parameter API"


This PR switches `shard_parameter` API to use `dist.scatter` instead of `dist.broadcast`. Instead of sending the whole tensor to each rank, we split the tensor beforehand, and only send the part needed to the corresponding rank, which greatly reduce the communication overhead.

Differential Revision: [D33933419](https://our.internmc.facebook.com/intern/diff/D33933419/)

[ghstack-poisoned]
